### ISSUE TITLE
mobject-store does not free underlying bake regions when removing an object

### ISSUE THREAD

**Issue by shanedsnyder** [mochi-hpc/mobject-test#18]
In GitLab by @shanedsnyder on Apr 4, 2018, 11:48

It should take advantage of the new `bake_remove()` calls inside of BAKE to free these regions and avoid leaking their memory.

**Comment by shanedsnyder** [comment]
In GitLab by @mdorier on Apr 4, 2018, 11:50

assigned to @mdorier

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on May 23, 2018, 11:07

Taking stock of what all needs to be cleaned up on a Mobject remove call. I think it is basically:

*  remove kv entries associating object segments to bake regions
*  remove any kv entries associated with omap entries
*  remove kv entries associating the object name with its OID (so Mobject can say that the object does not exist anymore)
*  remove the BAKE regions

Am I forgetting anything? The design doc mentions an optional map from OID back to object name, but not sure if that was implemented.

What about ordering of these operations? What's the proper order so the remove appears atomic from the user's perspective?

**Comment by shanedsnyder** [comment]
In GitLab by @mdorier on May 23, 2018, 11:12

You should start with the name => oid entry to make the object disappear from a client perspective.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on May 23, 2018, 11:16

OK, then I'm assuming everything else can be done asynchronously -- the object at that point will not exist to user's of the client API.

**Comment by shanedsnyder** [comment]
In GitLab by @carns on May 23, 2018, 18:54

You could consider having the mobject provider handler do this: 

- remove the critical key
- ack to client
- launch a detached ULT to do the rest

I'm assuming those last steps aren't very important semantically; you could think of it as doing a namespace remove and then kicking off a (very limited) background scrub.

PVFS did something similar by putting trove bstream removes in a queue to be removed by a dedicated thread, mainly because for a while there one of the popular Linux file systems (maybe ext3?) had a really slow unlink() operation and there wasn't much point in having client's wait around for it.  The server could die before completing the operation but that's always true even if you do the steps before the ack instead of after; you eventually need a background scrub to get that right.

**Comment by shanedsnyder** [comment]
In GitLab by @mdorier on May 24, 2018, 04:52

closed

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on May 24, 2018, 09:46

Sounds reasonable to me, I can take a stab at this, unless Matthieu is (hence the closing of the issue)?

**Comment by shanedsnyder** [comment]
In GitLab by @mdorier on May 24, 2018, 10:17

reopened

**Comment by shanedsnyder** [comment]
In GitLab by @mdorier on May 24, 2018, 10:18

Sorry I clicked "close" by mistake. If you have time to do it, be my guest. I have some things to finish that will take me at least the end of this week.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Jun 5, 2018, 16:45

I just submitted an initial cut at this in MR !6, and assigned to @mdorier to sanity check.

It should take care of everything we discussed, except everything is done in the foreground before ack'ing the client. I'll close out this issue after the MR is accepted and will open a new issue to track the optimization we discussed (bg thread for less critical cleanup tasks).
